adapters/http-vertx-base/src/test/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapterTest.java

```diff
@@ -410,6 +411,42 @@ public void testUploadCommandResponseFailsForDisabledTenant() {
         verify(metrics, never()).incrementCommandResponseDeliveredToApplication(eq("tenant"));
     }
 
+    /**
+     * Verifies that the adapter fails the upload of a command response with a 403 if the device is not registered and
+     * the valid credentials for the device is available.
+     */
+    @Test
+    public void testUploadCommandResponseFailsForDeletedDevices() {
+
+        final HttpServer server = getHttpServer(false);
+        final AbstractVertxBasedHttpProtocolAdapter<HttpProtocolAdapterProperties> adapter = getAdapter(server, null);
+        final Buffer payload = Buffer.buffer("some payload");
+        final RoutingContext ctx = newRoutingContext(payload, "application/text", mock(HttpServerRequest.class),
+                mock(HttpServerResponse.class));
+        final TenantObject to = TenantObject.from("tenant", true);
+
+        // Given an adapter that is enabled for a device's tenant
+        to.addAdapterConfiguration(new JsonObject()
+                .put(TenantConstants.FIELD_ADAPTERS_TYPE, ADAPTER_TYPE)
+                .put(TenantConstants.FIELD_ENABLED, true));
+        when(tenantClient.get(eq("tenant"), (SpanContext) any())).thenReturn(Future.succeededFuture(to));
+
+        // which is connected to a Credentials service that has credentials on record for device 9999
+        when(adapter.getAuthenticatedDevice(ctx)).thenReturn(new DeviceUser("tenant", "9999"));
+
+        // but for which no registration information is available
+        when(regClient.assertRegistration((String) any(), (String) any(), (SpanContext) any()))
+                .thenReturn(Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_FORBIDDEN,
```

Comment:
this should probably return a 404, reflecting the case that the device does not (or no longer) exist anymore, right?

Reply:
Device registration doesn't exist and where as device credentials exists. It seemed to me partially available and hence it led me to 403. I think so too, that 404 is more appropriate.